src/java/org/apache/cassandra/config/DataStorageSpec.java

```diff
@@ -38,117 +36,111 @@
     }
 
-    DataStorageSpec(long quantity, DataStorageUnit unit)
+    // This is used only for max_value_size for legacy reasons; we should be very careful about introducing similar constructors
+    // which do not validate for min unit
+    private DataStorageSpec(long quantity, DataStorageUnit unit, DataStorageUnit minUnit, long max)
     {
-        if (quantity < 0)
-            throw new ConfigurationException("Invalid data storage: value must be positive, but was " + quantity);
-
         this.quantity = quantity;
         this.unit = unit;
+
+        validateQuantity(quantity, unit, minUnit, max);
     }
 
-    public DataStorageSpec (String value, DataStorageUnit minUnit)
+    private DataStorageSpec(String value, DataStorageUnit minUnit)
     {
-        if (!MAP_UNITS_PER_MIN_UNIT.containsKey(minUnit))
-            throw new ConfigurationException("Invalid smallest unit set for " + value);
-
         //parse the string field value
-        Matcher matcher = STORAGE_UNITS_PATTERN.matcher(value);
+        Matcher matcher = UNITS_PATTERN.matcher(value);
 
         if (matcher.find())
         {
             quantity = Long.parseLong(matcher.group(1));
             unit = DataStorageUnit.fromSymbol(matcher.group(2));
 
-            if (!MAP_UNITS_PER_MIN_UNIT.get(minUnit).contains(unit))
-                throw new ConfigurationException("Invalid data storage: " + value + " Accepted units:" + MAP_UNITS_PER_MIN_UNIT);
+            //this constructor is used only by extended classes for smallest unit; upper bound is guarded there accordingly
+
+            validateMinUnit(unit, minUnit, value);
         }
         else
         {
-            throw new ConfigurationException("Invalid data storage: " + value + " Accepted units:" + MAP_UNITS_PER_MIN_UNIT.get(minUnit) +
+            throw new ConfigurationException("Invalid data storage: " + value + " Accepted units:" + acceptedUnits(minUnit) +
                                              " where case matters and only non-negative values are accepted");
         }
     }
-    /**
-     * Creates a {@code DataStorageSpec} of the specified amount of bytes.
-     *
-     * @param bytes the amount of bytes
-     * @return a {@code DataStorageSpec}
-     */
-    public static DataStorageSpec inBytes(long bytes)
+
+    private DataStorageSpec(String value, DataStorageUnit minUnit, long max)
     {
-        return new DataStorageSpec(bytes, DataStorageUnit.BYTES);
+        this(value, minUnit);
+
+        validateMinUnit(unit, minUnit, value);
+        validateQuantity(value, quantity(), unit(), minUnit, max);
     }
 
-    /**
-     * Creates a {@code DataStorageSpec} of the specified amount of kibibytes.
-     *
-     * @param kibibytes the amount of kibibytes
-     * @return a {@code DataStorageSpec}
-     */
-    public static DataStorageSpec inKibibytes(long kibibytes)
+    private void validateMinUnit(DataStorageUnit unit, DataStorageUnit minUnit, String value)
```

Comment:
Nit: this method could be `static`:
```suggestion
    private static void validateMinUnit(DataStorageUnit unit, DataStorageUnit minUnit, String value)
```